test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp

```diff
@@ -0,0 +1,182 @@
+Java_GetThreadStateMountedTest_testThread(
+  JNIEnv* jni, jclass clazz, jthread vthread, jboolean test_interrupt,
+  jint expected_strong, jint expected_weak)
+{
+  jint exp_ct_state = JVMTI_THREAD_STATE_ALIVE
+                      | JVMTI_THREAD_STATE_WAITING
+                      | JVMTI_THREAD_STATE_WAITING_INDEFINITELY;
+  jint exp_vt_state = expected_strong
+                      | JVMTI_THREAD_STATE_ALIVE;
+
+  jthread cthread = get_carrier_thread(jvmti, jni, vthread);
+
+  verify_thread_state("cthread", jni, cthread,
+                      exp_ct_state, 0);
+  verify_thread_state("vthread", jni, vthread,
+                      exp_vt_state, expected_weak);
+
+  // suspend ctread and verify
+  LOG("suspend cthread\n");
+  suspend_thread(jvmti, jni, cthread);
+  verify_thread_state("cthread", jni, cthread,
+                      exp_ct_state | JVMTI_THREAD_STATE_SUSPENDED, 0);
+  verify_thread_state("vthread", jni, vthread,
+                      exp_vt_state, expected_weak);
+
+  // suspend vtread and verify
+  LOG("suspend vthread\n");
+  suspend_thread(jvmti, jni, vthread);
+  verify_thread_state("cthread", jni, cthread,
+                      exp_ct_state | JVMTI_THREAD_STATE_SUSPENDED, 0);
+  verify_thread_state("vthread", jni, vthread,
+                      exp_vt_state | JVMTI_THREAD_STATE_SUSPENDED, expected_weak);
+
+  // resume cthread and verify
+  LOG("resume cthread\n");
+  resume_thread(jvmti, jni, cthread);
+  verify_thread_state("cthread", jni, cthread,
+                      exp_ct_state, 0);
+  verify_thread_state("vthread", jni, vthread,
+                      exp_vt_state | JVMTI_THREAD_STATE_SUSPENDED, expected_weak);
+
+  if (test_interrupt) {
+    // interrupt vthread (while it's suspended)
+    LOG("interrupt vthread\n");
+    check_jvmti_status(jni, jvmti->InterruptThread(vthread), "error in JVMTI InterruptThread");
+    verify_thread_state("cthread", jni, cthread,
+                        exp_ct_state, 0);
+    verify_thread_state("vthread", jni, vthread,
+                        exp_vt_state | JVMTI_THREAD_STATE_SUSPENDED | JVMTI_THREAD_STATE_INTERRUPTED,
+                        expected_weak);
+  }
+
+  // resume vthread;
+  LOG("resume vthread\n");
+  resume_thread(jvmti, jni, vthread);
+
+  // don't verify thread state after InterruptThread and ResumeThread
+}
+
+extern "C" JNIEXPORT int JNICALL
+Java_GetThreadStateMountedTest_getErrorCount(JNIEnv* jni, jclass clazz) {
+  return error_count;
+}
+
+
+static std::atomic<bool> time_to_exit(false);
+
+extern "C" JNIEXPORT void JNICALL
+Java_GetThreadStateMountedTest_waitInNative(JNIEnv* jni, jclass clazz) {
+  // Notify main thread that we are ready
+  jfieldID fid = jni->GetStaticFieldID(clazz, "waitInNativeReady", "Z");
+  if (fid == nullptr) {
+    jni->FatalError("cannot get waitInNativeReady field");
+    return;
+  }
+  jni->SetStaticBooleanField(clazz, fid, JNI_TRUE);
+
+  while (!time_to_exit) {
+    sleep_ms(100);
+  }
```

Comment:
Is this loop reliable in terms of thread states?

Reply:
The code is pure native (sleep_ms uses native functions, it does not interact with JVM), so there is no thread state changes